torch/csrc/dynamo/guards.cpp

```diff
@@ -822,6 +822,29 @@ class ID_MATCH : public LeafGuard {
   unsigned long long _expected;
 };
 
+class EQUALS_MATCH : public LeafGuard {
+ public:
+  EQUALS_MATCH(py::object value, py::object verbose_code_parts)
+      : LeafGuard(verbose_code_parts),
+        _value(value),
+        _value_type(Py_TYPE(value.ptr())) {}
+
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    return Py_TYPE(value) == _value_type &&
```

Comment:
I think some of the existing usage of this compare different types -- for example we compare torch.Size to tuple.

Reply:
There were 3 special cases here

1,2) numpy.isnan/math.isnan - Handled separately in the prototype [here](https://github.com/pytorch/pytorch/pull/110590/files#diff-61187d2fc8c5734622bcf785b2f864dca420e5033f9c152e413a4bcd8c143a5fR669-R671) by a separate LAMBDA_GUARD
3) torch.Size - Handled separately in the prototype [here](https://github.com/pytorch/pytorch/pull/110590/files#diff-9f0663783bcd93e948e0491ef61b48123bdc9977bcc632fd707da578df13bfa1R792)  by adding a TYPE_MATCH guard.

So, I decided to keep this one simple and rely on an additional guard for simplicity.